Extension Typing shim

DedSec256 opened the pull request:
We'd like to be able to instantiate and host type providers out-of-process. 
This will allow us to isolate the work of the IDE from the work of type providers and use runtime that is different to the one that hosts FCS.
This PR adds an extension typing shim that allows to override FCS type providers instantiating for the aforementioned purpose

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=8235) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=8235) before we can accept your contribution.<br/><br/>:x: DedSec256 [sign now](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=8235)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/fsharp?pullRequest=8235) it.</sub>

auduchinok commented:
There's a failing test `Test Project25 symbol uses of type-provided members` that looks related:
https://dev.azure.com/dnceng/public/_build/results?buildId=486778&view=logs&j=c3b5cd2e-ddad-540a-7ae1-df25e52fdba0&t=d4985d65-7d01-5e76-2879-f1b5d6535baa

DedSec256 commented:
> There's a failing test `Test Project25 symbol uses of type-provided members` that looks related:
> https://dev.azure.com/dnceng/public/_build/results?buildId=486778&view=logs&j=c3b5cd2e-ddad-540a-7ae1-df25e52fdba0&t=d4985d65-7d01-5e76-2879-f1b5d6535baa

![image](https://user-images.githubusercontent.com/26364714/72614737-17518800-3944-11ea-8708-f5a1261a2a9d.png)

Locally these tests are green.

cartermp commented:
Unfortunately we have some heisenbugs.

DedSec256 commented:
> Unfortunately we have some heisenbugs.

@cartermp, detected & fixed 😎 (thanks to @auduchinok)

dsyme commented:
Could you post an example of what out-of-proc hosting looks like?


This is still in-proc hosting, if I understand correctly? https://github.com/dotnet/fsharp/pull/8235/files#diff-f6b587f1e431a5977102d2a20c2a1990R29

DedSec256 commented:
> Could you post an example of what out-of-proc hosting looks like?
> 
> This is still in-proc hosting, if I understand correctly? https://github.com/dotnet/fsharp/pull/8235/files#diff-f6b587f1e431a5977102d2a20c2a1990R29

@dsyme, the shim is a part of out-of-process type providers protocol model which can be divided in two parts: the in-process one and the out-of-process one. ``InstantiateTypeProvidersOfAssembly(...)`` is called in the in-process area and in the current implementation it creates another process that uses the desired user .NET runtime, where real type providers entities are be created and live. It then create provided types proxies that live in the in-process area and replace the real types. These in-process provided types get info needed for FCS via custom protocol model.

![image](https://user-images.githubusercontent.com/26364714/76075377-f99cb880-5fad-11ea-95f7-7a9989a2ed2d.png)

DedSec256 commented:
https://github.com/dotnet/fsharp/pull/8235

DedSec256 commented:
- Added more necessary extension points
- Encapsulated work with Shim inside ExtensionTyping
- Removed Tainted as result type from InstantiateTypeProvidersOfAssembly

DedSec256 commented:
If you don't want to make Tainted API public, I can try refactoring the TaintedEquals/GetHashCode in such a way as to remove the Tainted <_> from the arguments and split ExntensionTyping module up into public (Shim and ProvidedModels only) part & internal one

dsyme commented:
> @dsyme, the shim is a part of out-of-process type providers protocol model which can be divided in two parts: the in-process one and the out-of-process one. 

So I'd like to see a bit more evidence that the proxying model works for this proposed extension point to the compiler.

Could you supply a dummy implementation of proxying as a test case as part of this PR?  

We kind of need evidence that the extension point fills the need it's designed to fill for out-of-proc cases

DedSec256 commented:
> Could you supply a dummy implementation of proxying as a test case as part of this PR?

Ok, I will prepare it and answer!

auduchinok commented on tests/service/ExtensionTypingProviderTests.fs:
```diff
@@ -15,13 +18,65 @@ open FSharp.Compiler.AbstractIL.Internal.Library
 open Microsoft.FSharp.Core.CompilerServices
 
+module internal XmlProxyProvidersProject = 
+    let fileName1 = Path.ChangeExtension(Path.GetTempFileName(), ".fs")
```
Could it use something like the following?
https://github.com/dotnet/fsharp/blob/ad112ad140717481d86e460fe8a646b5c4c17c47/tests/service/Common.fs#L170

DedSec256 commented:
@dsyme, I added a simple test demonstrating how extension points work within a single process. To transfer data about the provided models between processes, you must additionally make the provided models virtual to create your own proxy-provided-models implementations.

cartermp commented:
@dsyme could you take another look?

cartermp commented:
@dsyme could you take a look at this?

DedSec256 commented:
@cartermp, I will push the latest changes related to this PR shortly and open another PR closely related to the virtualization of provided entities (:

KevinRansom commented:
I'm closing this it looks like it's dead.  Please feel free to reopen with a comment, if you wish to pursue it.

Kevin

auduchinok commented:
@KevinRansom It still awaits a review by @dsyme AFAICT.

dsyme commented:
I will leave some review comments

However I really need to have a discussion with @vzarytovskii , @KevinRansom and @TIHan about whether this is the hook we want. Notes below

- The PR adds a global hook into FCS to intercept all type provider instantiations, with the intention to use this to run the type providers out of process.  
- The PR has testing, but doesn't actually demonstrate that this allows TPs to really, truly be run out of process

On the whole I'm sympathetic to this.  Type providers are an extension point that can cause arbitrary instability and faults and it is reasonable to have a general mechanism to run them outside the main checking process (even if that itself is outside the IDE) without requiring the modification of each type provider.  It is a real problem that needs solving.

However it feels like we really need evidence that this actually works in practice, and that the architecture doesn't need to be changed further to enable effective out-of-proc.

auduchinok commented:
> However it feels like we really need evidence that this actually works in practice

@dsyme Due the runtime that Rider uses to host FCS is often different to the runtime used by the compiler invoked by MSBuild, there were many issues about various type providers not working in the IDE. A combination of this PR and https://github.com/dotnet/fsharp/pull/11110 has been working in Rider since the previous release and the most of the issues we've encountered before were fixed with this shim. Here're examples of the issues fixed: [RIDER-31245](https://youtrack.jetbrains.com/issue/RIDER-31245), [RIDER-38184](https://youtrack.jetbrains.com/issue/RIDER-38184), [RIDER-54377](https://youtrack.jetbrains.com/issue/RIDER-54377).

In addition to verifying the fix by hand, there're tests covering some popular and/or interesting for testing providers too: https://github.com/JetBrains/resharper-fsharp/tree/net212/rider-fsharp/src/test/kotlin/typeProviders

> the architecture doesn't need to be changed further to enable effective out-of-proc.

We'll be happy to help evolving it in future if needed. Having that said, having it merged upstream will simplify maintaining a lot, since if there're some changes that break this approach, we'll have to come here and create an issue and work on a fix, likely involving the F# team in some way, anyway.

dsyme commented on src/fsharp/ExtensionTyping.fs:
```diff
@@ -751,9 +745,108 @@ module internal ExtensionTyping =
+    module Shim =
+
+        type IExtensionTypingProvider =
+            abstract InstantiateTypeProvidersOfAssembly:
```
Please put all these arguments into a single object called `ExtensionTypingProviderContext`

dsyme commented on src/fsharp/ExtensionTyping.fs:
```diff
@@ -751,9 +745,108 @@ module internal ExtensionTyping =
+                    if fullName then tp.GetType().FullName else tp.GetType().Name
+
+        let mutable ExtensionTypingProvider = DefaultExtensionTypingProvider() :> IExtensionTypingProvider
+        let shimLogger (tpe: TypeProviderError) =
```
This should either be internal or named more apprioriately

dsyme commented on src/fsharp/ExtensionTyping.fsi:
```diff
@@ -12,7 +12,7 @@ open FSharp.Core.CompilerServices
 open FSharp.Compiler.AbstractIL.IL
 open FSharp.Compiler.Text
 
-module internal ExtensionTyping =
```
This should not be public.  Please update the baselines for FCS surface area so we could see properly what's been added

dsyme commented on src/fsharp/tainted.fs:
```diff
@@ -77,19 +77,19 @@ type internal TypeProviderError
+type TaintedContext = { TypeProvider : ITypeProvider; TypeProviderAssemblyRef : ILScopeRef; TypeProviderDesignation: string; Lock : TypeProviderLock }
 
-[<NoEquality>][<NoComparison>] 
-type internal Tainted<'T> (context: TaintedContext, value: 'T) =
```
I don't believe we want `Tainted` to be public

DedSec256 replied:
Tainted is currently used in TaintedEquals methods in provided entities. If we transfer this logic, for example, inside Tainted, then we could make Tainted internal back

dsyme commented on src/fsharp/ExtensionTyping.fsi:
```diff
@@ -337,4 +340,33 @@ module internal ExtensionTyping =
+    [<AutoOpen>]
+    module Shim =
+
+        type IExtensionTypingProvider =
```
Please add `///` comments to everything made public

dsyme commented on src/fsharp/ExtensionTyping.fsi:
```diff
@@ -337,4 +340,33 @@ module internal ExtensionTyping =
+        type IExtensionTypingProvider =
+
+             /// Find and instantiate the set of ITypeProvider components for the given assembly reference
+            abstract InstantiateTypeProvidersOfAssembly :
```
Please give names and `///` comments to all parameters in a public API

dsyme commented on src/fsharp/tainted.fs:
```diff
@@ -77,19 +77,19 @@ type internal TypeProviderError
                 f (new TypeProviderError(errNum, tpDesignation, m, [msg], typeNameContext, methodNameContext))
 
-type TaintedContext = { TypeProvider: ITypeProvider; TypeProviderAssemblyRef: ILScopeRef; Lock: TypeProviderLock }
+type TaintedContext = { TypeProvider : ITypeProvider; TypeProviderAssemblyRef : ILScopeRef; TypeProviderDesignation: string; Lock : TypeProviderLock }
```
No space before `:` please

dsyme commented on src/fsharp/tainted.fs:
```diff
@@ -77,19 +77,19 @@ type internal TypeProviderError
-[<NoEquality>][<NoComparison>] 
-type internal Tainted<'T> (context: TaintedContext, value: 'T) =
+[<NoEquality>][<NoComparison>]
+type Tainted<'T> (context : TaintedContext, value : 'T) =
```
No space before `:` please

dsyme reviewed (changes requested):
I left a few comments.  It feels like too much is being made public.  Can we have baseline updates too please?

auduchinok commented:
@DedSec256 Could you address the issues please?

dsyme commented:
I think this can be closed now.  I just can't see us exposing this extension point from FCS - if we were to expose anything it would be a `ITypeProvider -> ITypeProvider` intercept that wraps a function around all type providers, allowing the FCS host to do whatever it likes with any `ITypeProvider`.  This doesn't need all these Provided* types to be exposed.

DedSec256 commented:
We have some ideas for this. I think we'll be back soon (:
